File: python/arrays_aslist.py
# check: https://rules.sonarsource.com/java/RSPEC-3631
import utils

import bblfsh

client = bblfsh.BblfshClient("0.0.0.0:9432")

uast = client.parse("../java/arrays_aslist.java").uast
calls = bblfsh.filter(uast, "//MethodInvocation/"
        "Identifier[@roleCall and @roleReceiver and @Name='Arrays']/parent::MethodInvocation/"
        "Identifier[@roleCall and @roleCallee and @Name='asList']/parent::MethodInvocation")

for c in calls:
    child_args = bblfsh.filter(c, "//*[@roleArgument and @roleLiteral]")
    if len(list(child_args)):
        print("Don't use slow Arrays.asList with primitives, line: {}"
                .format(c.start_position.line))

File: java/arrays_aslist.java
class ArraysAsList {
    void test() {
        Arrays.asList("a1", "a2", "b1", "c2", "c1"); // this is fine (Object)
        Arrays.asList(1, 2, 3, 4);
        Arrays.asList(01, 02);
        Arrays.asList(0x1, 0x2);
        Arrays.asList(0b11, 0b10);
        Arrays.asList(123.456);
        Arrays.asList('a', 'b');
        Arrays.asList(true, false);
    }
}
